From 2e28952e571a3ca5f5c3f03840f5047b35c30e53 Mon Sep 17 00:00:00 2001
From: Wen He <wen.he_1@nxp.com>
Date: Fri, 15 Nov 2019 16:53:06 +0800
Subject: [PATCH 06/30] drm/arm/malidp: Fix NULL pointer access for
 malidp_se_check_scaling

commit bfa77f0437a83910284ff837c2235b18d4073b33 from
http://source.codeaurora.org/external/qoriq/qoriq-components/linux

Fixed the crash issue caused by NULL pointer dereference when built
with GCC 9.2.

Crash information:
Unable to handle kernel NULL pointer dereference at virtual address
0000035c
Mem abort info:
Exception class = DABT (current EL), IL = 32 bits
SET = 0, FnV = 0
EA = 0, S1PTW = 0
Data abort info:
ISV = 0, ISS = 0x00000004
CM = 0, WnR = 0
user address but active_mm is swapper
Internal error: Oops: 96000004 [#1] PREEMPT SMP
Modules linked in:
Process swapper/0 (pid: 1, stack limit = 0xffff000008038000)
CPU: 0 PID: 1 Comm: swapper/0 Not tainted 4.14.140-00001-ge082b5def026 #1
Hardware name: LS1028A RDB Board (DT)
task: ffff80207b346d00 task.stack: ffff000008038000
PC is at malidp_de_plane_check+0x280/0x2b8
LR is at malidp_de_plane_check+0x1a4/0x2b8

Signed-off-by: Wen He <wen.he_1@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/gpu/drm/arm/malidp_planes.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/gpu/drm/arm/malidp_planes.c b/drivers/gpu/drm/arm/malidp_planes.c
index 29409a65d864..7a4e9a03831c 100644
--- a/drivers/gpu/drm/arm/malidp_planes.c
+++ b/drivers/gpu/drm/arm/malidp_planes.c
@@ -162,6 +162,7 @@ static int malidp_se_check_scaling(struct malidp_plane *mp,
 		src_h = state->src_h >> 16;
 	}
 
+	mc = to_malidp_crtc_state(crtc_state);
 	if ((state->crtc_w == src_w) && (state->crtc_h == src_h)) {
 		/* Scaling not necessary for this plane. */
 		mc->scaled_planes_mask &= ~(mp->layer->id);
-- 
2.17.1

